Dynatrace/dynatrace-operator
# Rework how istio is handled

## Description
So the "old" way (what is currently in main) how we handle istio is described in this image: (press the `<-->` looking icon to view it in big 😅  )
```mermaid
---
title: Old
---
flowchart LR
    dynakube[Dynakube\nName: test\nEnableIstio: true]
    subgraph operator
        direction TB
        reconcile-everything...
        reconcile-istio
    end
    subgraph for-api-url
        direction TB
        se1[ServiceEntry\nName: some-hash \nPurpose: For the api-url]
        vs1[VirtualService\nName: some-hash \nPurpose: For the api-url]
        se1 -.- vs1
    end
    subgraph for-communication-hosts
        direction TB
        se2[ServiceEntry\nName: some-hash \nPurpose: For oneagent IP-based communication-hosts]
        se3[ServiceEntry\nName: some-hash \nPurpose: For oneagent FQDN-based communication-hosts]
        vs2[VirtualService\nName: some-hash \nPurpose: For oneagent FQDN-based communication-hosts]
        se3 -.- vs2
        se2 ~~~ se3
    end


    exp1([Redundant, as for the operator to get to this point, this should already exist]):::bad-red
    exp2([+there is a overcomplicated delete logic, as the names constantly change, because it's hash based on the contents\n+If any error happens, we just log it, Which can silently break functionality]):::bad-red

    dynakube-->operator
    reconcile-istio-->for-api-url
    reconcile-istio-->for-communication-hosts
    for-api-url --> exp1
    for-communication-hosts --> exp2
    exp1-->exp2

    classDef bad-red stroke:#f00
```
According to this PR how it works:
```mermaid
---
title: New
---
flowchart LR
    dynakube[Dynakube\nName: test\nEnableIstio: true]
    subgraph operator
        direction TB
        reconcile-istio-for-api-url
        reconcile-connectionInfo...
        reconcile-istio-communication-hosts
        reconcile-everything-else...
    end
    subgraph for-api-url
        direction TB
        se1[ServiceEntry\nName: test-dk-fqdn-operator \nPurpose: For the api-url]
        vs1[VirtualService\nName: test-dk-fqdn-operator \nPurpose: For the api-url]
        se1 -.- vs1
    end
    subgraph for-communication-hosts
        direction TB
        se2[ServiceEntry\nName: test-dk-ip-oneagent \nPurpose: For oneagent IP-based communication-hosts]
        se3[ServiceEntry\nName: test-dk-fqdn-oneagent \nPurpose: For oneagent FQDN-based communication-hosts]
        vs2[VirtualService\nName: test-dk-fqdn-oneagent \nPurpose: For oneagent FQDN-based communication-hosts]
        se3 -.- vs2
        se2 ~~~ se3
    end


    exp1([No longer redundant, user no longer need to configure istio for the API Url by hand]):::good-green
    exp2([Update logic is simple as the names are static + names mean something\nincase of error, we actually stop what we are doing]):::good-green

    dynakube-->operator
    reconcile-istio-for-api-url-->for-api-url
    reconcile-istio-communication-hosts-->for-communication-hosts
    for-api-url --> exp1
    for-api-url --> exp2
    for-communication-hosts --> exp2

    classDef good-green stroke:#0f0
```

TLDR:
How we used to handle the istio integration was less then ideal:
- The user had to do extra steps for the operator to even get to the isito reconcile part
- The resources we created had a hash for a name, that only complicated things and confused users
- The resources we created had weird labels (that we were afraid to change as they were used as labels selector due to the names being unpredictable)
- The code of the istio reconciler was "refactored" multiple times, but was never re-thinked, so the code itself was overly convoluted for what it was doing.

~Still TODO:~
~- fix e2e tests...~
The fix for the e2e test was rather simple as its not the best and that should be fixed separately in a follow up. (also there are multiple e2e PRs in the work so I don't want to add to the pile right now)

## How can this be tested?

1. have a cluster with istio with `REGISTRY_ONLY` mode
2. have istio inject the operator's namespace
3. deploy a dynakube with `EnableIstio: true`
4. it should just work
    - previously you would need to create a `ServiceEntry` + `VirtualService` for your APIURL to get to the istio reconcile part, you DON'T have to do this anymore
    -  The created `ServiceEntries` + `VirtualServices` should be named nicely and have nice labels.


## Checklist
